File: sdks/python/src/echo_agent_sdk/subagent_usage_values.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True)
class LlmUsageStats:
    model: str = ""
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    cached_prompt_tokens: int = 0
    cache_creation_prompt_tokens: int = 0
    usage_reported: bool = False
    call_count: int = 0

    def record(
        self,
        model: str,
        prompt_tokens: int,
        completion_tokens: int,
        total_tokens: int,
        cached_prompt_tokens: int,
        cache_creation_prompt_tokens: int,
        usage_reported: bool,
    ) -> None:
        self.model = model
        self.prompt_tokens += prompt_tokens
        self.completion_tokens += completion_tokens
        self.total_tokens += total_tokens
        self.cached_prompt_tokens += cached_prompt_tokens
        self.cache_creation_prompt_tokens += cache_creation_prompt_tokens
        self.usage_reported = self.usage_reported or usage_reported
        self.call_count += 1

    def to_payload(self, session_id: str) -> dict[str, str | bool | int]:
        return {
            "session_id": session_id,
            "model": self.model or "unknown",
            "prompt_tokens": self.prompt_tokens,
            "completion_tokens": self.completion_tokens,
            "total_tokens": self.total_tokens,
            "cached_prompt_tokens": self.cached_prompt_tokens,
            "cache_creation_prompt_tokens": self.cache_creation_prompt_tokens,
            "usage_reported": self.usage_reported,
            "call_count": self.call_count,
        }

File: sdks/typescript/src/subagent_usage_values.ts
/** Cumulative Subagent LLM usage values without owning provider execution. */
export class LlmUsageStats {
  public model = "";
  public promptTokens = 0n;
  public completionTokens = 0n;
  public totalTokens = 0n;
  public cachedPromptTokens = 0n;
  public cacheCreationPromptTokens = 0n;
  public usageReported = false;
  public callCount = 0n;

  public record(model: string, promptTokens: bigint, completionTokens: bigint, totalTokens: bigint,
    cachedPromptTokens: bigint, cacheCreationPromptTokens: bigint, usageReported: boolean): void {
    this.model = model;
    this.promptTokens += promptTokens;
    this.completionTokens += completionTokens;
    this.totalTokens += totalTokens;
    this.cachedPromptTokens += cachedPromptTokens;
    this.cacheCreationPromptTokens += cacheCreationPromptTokens;
    this.usageReported ||= usageReported;
    this.callCount += 1n;
  }

  public toPayload(sessionId: string): Record<string, string | boolean> {
    return {
      session_id: sessionId,
      model: this.model || "unknown",
      prompt_tokens: this.promptTokens.toString(),
      completion_tokens: this.completionTokens.toString(),
      total_tokens: this.totalTokens.toString(),
      cached_prompt_tokens: this.cachedPromptTokens.toString(),
      cache_creation_prompt_tokens: this.cacheCreationPromptTokens.toString(),
      usage_reported: this.usageReported,
      call_count: this.callCount.toString(),
    };
  }
}
